File: java/direct-policy/src/test/java/org/nhindirect/policy/impl/BinaryCollectionPolicyOperatorExecutor_executeTest.java
package org.nhindirect.policy.impl;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;

import org.nhindirect.policy.PolicyOperator;
import org.nhindirect.policy.PolicyValue;
import org.nhindirect.policy.PolicyValueFactory;

import junit.framework.TestCase;

public class BinaryCollectionPolicyOperatorExecutor_executeTest extends TestCase
{
	public void testExecute_intersection_assertResults()
	{
		// both intersection of 1,2,3
		PolicyValue<List<String>> op1 = PolicyValueFactory.getInstance(Arrays.asList("A", "B", "C", "D"));
		PolicyValue<List<String>> op2 = PolicyValueFactory.getInstance(Arrays.asList("A", "B", "E"));
		
		BinaryCollectionPolicyOperatorExecutor<List<String>,List<String>> operator = 
				new BinaryCollectionPolicyOperatorExecutor<List<String>,List<String>>(op1, op2, PolicyOperator.INTERSECTION);
		
		Collection<?> resultList = operator.execute().getPolicyValue();
		assertEquals(2, resultList.size());
		assertTrue(resultList.contains("A"));
		assertTrue(resultList.contains("B"));
		

		op1 = PolicyValueFactory.getInstance(Arrays.asList("A", "B", "C"));
		
		op2 = PolicyValueFactory.getInstance(Arrays.asList("D", "E", "F"));
		
		operator = new BinaryCollectionPolicyOperatorExecutor<List<String>,List<String>>(op1, op2, PolicyOperator.INTERSECTION);
		assertTrue(operator.execute().getPolicyValue().isEmpty());

	}
}

File: csharp/unittests/policy.tests/BinaryCollectionPolicyOperatorExecutor_executeTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using FluentAssertions;
using Xunit;

namespace Health.Direct.Policy.Tests
{
    public class BinaryCollectionPolicyOperatorExecutor_executeTest
    {
        [Fact]
        public void TestExecute_Intersection_AssertResults()
        {
            // both intersection of 1,2,3

            IPolicyValue<IList<String>> op1 = PolicyValueFactory<IList<String>>.GetInstance(new List<string> { "A", "B", "C", "D" });
            IPolicyValue<IList<String>> op2 = PolicyValueFactory<IList<String>>.GetInstance(new List<string> { "A", "B", "E" });

            IEnumerable<string> resultList = PolicyOperator<IList<string>, IList<string>, IEnumerable<string>>.INTERSECT.Execute(op1.GetPolicyValue(), op2.GetPolicyValue());

            var result = resultList.ToList();
            result.Count().Should().Be(2);
            result.Contains("A").Should().BeTrue();
            result.Contains("B").Should().BeTrue();
            
            op1 = PolicyValueFactory<IList<String>>.GetInstance(new List<string> { "A", "B", "C" });
            op2 = PolicyValueFactory<IList<String>>.GetInstance(new List<string> { "D", "E", "F" });

            resultList = PolicyOperator<IList<string>, IList<string>, IEnumerable<string>>.INTERSECT.Execute(op1.GetPolicyValue(), op2.GetPolicyValue());
            resultList.Any().Should().BeFalse();

            //
            // Change ordering
            //
            op1 = PolicyValueFactory<IList<String>>.GetInstance(new List<string> { "D", "C", "B", "A" });
            op2 = PolicyValueFactory<IList<String>>.GetInstance(new List<string> { "A", "E", "B" });
            resultList = PolicyOperator<IList<string>, IList<string>, IEnumerable<string>>.INTERSECT.Execute(op1.GetPolicyValue(), op2.GetPolicyValue());

            result = resultList.ToList();
            result.Count().Should().Be(2);
            result.Contains("A").Should().BeTrue();
            result.Contains("B").Should().BeTrue();
        }
    }
}
